src/main/java/frc/team2412/robot/util/auto/AutoLogic.java

```diff
@@ -356,4 +332,176 @@ public static double getEstimatedAutoDuration() {
 		}
 		return 0;
 	}
+
+	// commands util
+
+	public static BooleanSupplier isReadyToLaunch() {
+		// return () -> (s.launcherSubsystem.isAtAngle() && s.launcherSubsystem.isAtSpeed());
+
+		// Should be able to launch if:
+		// Launcher is at the correct angle and flywheel rpm
+		// Note is finished indexing (intake all in command is finished)
+		return (INTAKE_ENABLED & LAUNCHER_ENABLED
+				? () ->
+						(s.launcherSubsystem.isAtAngle()
+								&& s.launcherSubsystem.isAtSpeed()
+								&& s.intakeSubsystem.getCurrentCommand() == null)
+				: () -> true);
+	}
+
+	public static BooleanSupplier untilNoNote() {
+		// decided to go from checking for note in feeder to both feeder and index in case note is still
+		// indexing
+		return (INTAKE_ENABLED
+				? () -> !(s.intakeSubsystem.feederSensorHasNote() && s.intakeSubsystem.indexSensorHasNote())
+				: () -> true);
+	}
+
+	// Should be able to tell if a robot has a note based off if intake is still running when checked,
+	// since if note is being indexed, intake motors should've been disabled.
+	public static BooleanSupplier hasNoNote() {
+		return (INTAKE_ENABLED ? () -> !(s.intakeSubsystem.isIntakeRunning()) : () -> true);
+	}
+
+	// registered commands
+
+	public static Command subwooferLaunch() {
+
+		// return (LAUNCHER_ENABLED && INTAKE_ENABLED && APRILTAGS_ENABLED
+		// 		? stopFeeder()
+		// 				.andThen(
+		// 						Commands.either(Commands.none(), index(), s.intakeSubsystem::feederSensorHasNote))
+		// 				.andThen(
+		// 						new SetAngleLaunchCommand(
+		// 										s.launcherSubsystem,
+		// 										LauncherSubsystem.SPEAKER_SHOOT_SPEED_RPM,
+		// 										LauncherSubsystem.SUBWOOFER_AIM_ANGLE)
+		// 								.until(isReadyToLaunch())
+		// 								.andThen(new WaitCommand(FEEDER_DELAY))
+		// 								.andThen(new FeederInCommand(s.intakeSubsystem).until(untilNoNote())))
+		// 		: Commands.none());
+
+		// Checks if the robot has a note in the subsystem, if it does, launch
+		return (LAUNCHER_ENABLED && INTAKE_ENABLED && APRILTAGS_ENABLED
+				? stopFeeder()
+						.andThen(
+								Commands.either(
+										new SetAngleLaunchCommand(
+														s.launcherSubsystem,
+														LauncherSubsystem.SPEAKER_SHOOT_SPEED_RPM,
+														LauncherSubsystem.SUBWOOFER_AIM_ANGLE)
+												.until(isReadyToLaunch())
+												.andThen(new WaitCommand(FEEDER_DELAY))
+												.andThen(new FeederInCommand(s.intakeSubsystem).until(untilNoNote()))
+												.andThen(new WaitCommand(0.4))
+												.withName("Auto - SubwooferLaunchCommand"),
```

Comment:
For these and the other command compositions, we should call `withName()` on the outermost command, and not a command inside the composition, as the outermost command will be the one that is logged.